### DISCUSSION TITLE
Help wanted: Eleventy Plugin Directus

### DISCUSSION THREAD

**Discussion by lexoyo** [directus/directus#17293] [Archive]
Hello

I created a [Directus plugin for 11ty static site generator](https://github.com/silexlabs/eleventy-plugin-directus/)

It is useful to expose Directus collections as [global data in 11ty](https://www.11ty.dev/docs/data-global/) to access in 11ty templates and creates filters to manage assets URLs and translations. This plugin is also an important piece of the next version of [Silex website builder](https://www.silexlabs.org/silex-v3-kickoff/)

Would someone be kind enough to review my code and help me make it better? Here is where I retrieve directus data, which is a little hacky, as I use the `?fields=*.*.*` param of the API: https://github.com/silexlabs/eleventy-plugin-directus/blob/main/_scripts/client.js#L8

To your good heart directus community :pray:

**Comment by rijkvanzanten** [comment]
Does this download all the data from Directus first, and then render out the templates, or is the data fetched when the templates are rendered? 🤔 Like you correctly stated, the `*.*.*.*.*.*.*` usage of `fields` is a bit of a no-no, as you end up loading a lot more data than you need, and possibly exponentially increasing the size if you have a recursive relationship (like `articles->related_articles->articles`).

I'm curious if you could potentially extract what fields are used in buildtime from the template, and then dynamically only load the data that's used in the template on the fly 🤔

**Reply by lexoyo** [reply]
Thank you for your answer !

> Does this download all the data from Directus first, and then render out the templates, or is the data fetched when the templates are rendered? 

the usual way is to use data files (js) which do the query only once when queried by the template, e.g. [this example with contentful](https://github.com/contentful/11ty-contentful-starter/blob/master/_data/contentful-page.js)

> thinking Like you correctly stated, the `*.*.*.*.*.*.*` usage of `fields` is a bit of a no-no, as you end up loading a lot more data than you need, and possibly exponentially increasing the size if you have a recursive relationship (like `articles->related_articles->articles`).

yes that is definitely temporary, I need a better way

> I'm curious if you could potentially extract what fields are used in buildtime from the template, and then dynamically only load the data that's used in the template on the fly thinking

This could be achieved with the same technique, a "page.js" data files which retrieve pages with the exact structure

Or maybe with a shortcode, e.g. `{{ directus: "page", "*.*,images.thumbnails.*" }}`, that way we would query only what we need? But then depending on the template we would potentially download some of the data over and over, e.g. if we different fields but also some of the same fields

Or there could be a shortcode to get the "*" fields and then a filter to query other fields: `{{ directus.get: "page" | withFields: ["images"] }}`

I am willing to make this plugin better, even if I am a little worried that it will make a lot of calls to directus, especially when I have sites with hundreds of products, with thousands of pictures in different languages

Maybe I should find a clean way to recursively downloading a collection as a big json including the different linked data (m2a, m2m, o2m, files...)?

**Reply by rijkvanzanten** [reply]
> > Does this download all the data from Directus first, and then render out the templates, or is the data fetched when the templates are rendered?

> the usual way is to use data files (js) which do the query only once when queried by the template, e.g. [this example with contentful](https://github.com/contentful/11ty-contentful-starter/blob/master/_data/contentful-page.js)

Yeah! However, I think that the real power and beauty of your current plugin is the fact that you can use dot notation in the templates itself to access the nested relational content (eg `{{ post.author.first_name }}`), but in order to allow that, you have to fetch that nested content somehow. This is where we're currently in a bit of a pickle, as using the "simple" approach to just fetch everything up front means you have to rely on the `*.*.*` hackyness, which isn't ideal (although it works!)

> > I'm curious if you could potentially extract what fields are used in buildtime from the template, and then dynamically only load the data that's used in the template on the fly thinking
> 
> This could be achieved with the same technique, a "page.js" data files which retrieve pages with the exact structure
> 
> Or maybe with a shortcode, e.g. `{{ directus: "page", "*.*,images.thumbnails.*" }}`, that way we would query only what we need? But then depending on the template we would potentially download some of the data over and over, e.g. if we different fields but also some of the same fields
> 
> Or there could be a shortcode to get the "*" fields and then a filter to query other fields: `{{ directus.get: "page" | withFields: ["images"] }}`
> 
> I am willing to make this plugin better, even if I am a little worried that it will make a lot of calls to directus, especially when I have sites with hundreds of products, with thousands of pictures in different languages
> 
> Maybe I should find a clean way to recursively downloading a collection as a big json including the different linked data (m2a, m2m, o2m, files...)?

This is where it gets interesting (and very complicated!). If you were to fetch the data separately for each of the template uses of Directus, I agree that you'd probably end up DDoSing your instance, and slowing down things considerably. I'm curious if there's a way to do things in effectively two passes. The first path collects all the data that's accessed in the template first, so we can make a single (or the least amount of) call(s) to get all that data, and then the second pass "just" renders the templates with the data we've now fetched already. That way, the end user can do any level of nesting or other interesting data uses, while the plugin still limits the data fetching to the minimum level of loading from the API. 🤔

**Reply by lexoyo** [reply]
Hmm really interesting idea, best of both worlds

Maybe with filters but the syntax would be strange (`{{ "item.property" | directus }}`. Then maybe with async filters it would continue to parse the template. But I am not sure it would and more importantly it would be challenging to to build a model of all the required fields as `item` would be a variable in a for loop... Am I making sense? Is it even possible ?

The syntax could also be `{{ item | directus: "propertyObj.propertyStr" }}` but same difficulty to make a `?fields=...` out of it ?

**Reply by lexoyo** [reply]
Well it might be possible, I should test if we can get a 2 pass filter thanks to async filters
https://www.11ty.dev/docs/filters/#asynchronous-universal-filters

**Reply by lexoyo** [reply]
The solutions i see

1. Current solution : download everything, "deeply" enough - hacky inefficient, easy to use in templates
1. The 2 pass idea, figure out what needs to be downloaded out of templates - hacky and potentially inefficient on 11ty's side, easy to use too
1. Define in the plugin config what data should be downloaded, like TinaCMS does ([doc](https://tina.io/docs/schema/)), could be linked to directus yaml model, could probably be used in a graphql query directly - efficient and clean, requires to describe content one more time

Maybe solution 1 could also be optimized simply by omitting some useless data, e.g anything which is not a field like the "owner" collection in a m2m or o2m relationship

**Reply by rijkvanzanten** [reply]
Solution 1 is probably good enough for a lot of use cases, though number 2 would be ideal (but is way more difficult to build!)

**Reply by lexoyo** [reply]
I'll document this in the readme and see what people need then :)

Thank you for your precious input !

**Comment by lexoyo** [comment]
Hello again

> TL;DR: I want to rewrite the plugin clean, I don't manage to lookup a field with the rest API

I took time to think about it all and I have a better idea than the hack of fields:*.*.*.*

Here is what the plugin could be:

* Expose to eleventy templates all directus collections like this `directus.collections.posts` or `directus.collections.pages`
* For simple types one could do `{{ directus.collections.posts[0] | get: "title" }}`
* For relations (m2m, o2m, m2a) provide a filter which looks up the value(s) in the related table, e.g. `{% assign author = directus.collections.posts[0] | get: "author" %}`
* Handle singletons the same way
* Provide a translation filter to find a field of type translations and loop over it to find the right translation object: `{% assign translated = directus.collections.my_collection | translate: "fr" %}`

I think it is possible to do this in a clean way and it is much nicer than the official way (using 1 data file per collection, hard coding the model in the data files...)

**Here is my question:** 

Let's say I have the collection `posts` with a filed `related_pages`, then I want `{{ directus.collections.posts[1] | get: "related_pages" }}` to return a list of pages. Is there a way to know that the IDs in "related_posts" are from a collection "pages" ?

I tried these API calls but it seems impossible to know it just with the collection name "posts" + field name "related_pages" 

GET /items/posts/1
```json
{"data":{"id":3,"title":"test","related_pages":[1,2]}}
```
GET /fields/posts/related_pages
```json
{"data":{"collection":"posts","field":"related_pages","type":"alias","meta":{"id":916,"collection":"posts","field":"related_pages","special":["m2m"],"interface":"list-m2m","options":null,"display":null,"display_options":null,"readonly":false,"hidden":false,"sort":null,"width":"full","translations":null,"note":null,"conditions":null,"required":false,"group":null,"validation":null,"validation_message":null},"schema":null}}
```

**Reply by rijkvanzanten** [reply]
> I tried these API calls but it seems impossible to know it just with the collection name "posts" + field name "related_pages"

Have you tried the `/relations` endpoint? That should contain the data you're looking for!

**Reply by lexoyo** [reply]
Thank you for this pointer, sorry I didn't see it by myself

It would help a lot if you tell me if the algorithm bellow is how I am supposed to find the list of related_pages?

1. A template calls `get({"collection":"posts","title":"test","related_pages":[1]}, "related_pages")`
1. So I am looking for `related_pages` on `posts` with **`id == 1`**
1. I lookup `/relations/` for `meta.one_collection == "posts" && meta.one_field == "related_pages"` which gives me
    ```json
    {
      "collection": "posts_page",
      "field": "posts_id",
      "related_collection": "posts",
      "schema": {
        "table": "posts_page",
        "column": "posts_id",
        "foreign_key_table": "posts",
        "foreign_key_column": "id",
        "constraint_name": "posts_page_posts_id_foreign",
        "on_update": "NO ACTION",
        "on_delete": "SET NULL"
      },
      "meta": {
        "id": 217,
        "many_collection": "posts_page",
        "many_field": "posts_id",
        "one_collection": "posts",
        "one_field": "related_pages",
        "one_collection_field": null,
        "one_allowed_collections": null,
        "junction_field": "page_id",
        "sort_field": null,
        "one_deselect_action": "nullify"
      }
    }
    ``` 
1. I use `meta.many_collection` which value is `posts_page` to get the `id == 1` in `/items/posts_page/` which gives me:
    ```json
    {
      "id": 1,
      "posts_id": 3,
      "page_id": 18
    },
    ```
1. I use `meta.junction_field` which value is `page_id` to get `"page_id": 3`
1. I lookup `"field": "page_id" && "collection": "posts_page"` from `/relations/` which gives:
   ```json
    {
      "collection": "posts_page",
      "field": "page_id",
      "related_collection": "page",
      "schema": {
        "table": "posts_page",
        "column": "page_id",
        "foreign_key_table": "page",
        "foreign_key_column": "id",
        "constraint_name": "posts_page_page_id_foreign",
        "on_update": "NO ACTION",
        "on_delete": "SET NULL"
      },
      "meta": {
        "id": 216,
        "many_collection": "posts_page",
        "many_field": "page_id",
        "one_collection": "page",
        "one_field": null,
        "one_collection_field": null,
        "one_allowed_collections": null,
        "junction_field": "posts_id",
        "sort_field": null,
        "one_deselect_action": "nullify"
      }
    },
   ```
1. I use `"foreign_key_table": "page"` and `"foreign_key_column": "id"` from `schema` to get from `page` the item `id == 18`, which is what I was looking for

Is that what I am supposed to do?

Thank you so much for your help

**Reply by lexoyo** [reply]
Short version:

1. `get({"collection":"posts","title":"test","related_pages":[1]}, "related_pages")`
2. get item from `/relations/` where meta.one_collection == "posts" && meta.one_field == "related_pages" => use `meta.many_collection` to get the collection `posts_page`
3. get item from `/items/posts_page/` where `id == 1` => use `meta.junction_field` which value is `page_id` to get `"page_id": 3`
5. get item from `/relations/` where `"field": "page_id" && "collection": "posts_page"` => use "schema.foreign_key_table": "page" and "schema.foreign_key_column": "id" from `schema` to get the item from page where `id == 18`

**Reply by rijkvanzanten** [reply]
> Is that what I am supposed to do?

That'd be one way, but it would be easier to rely on nesting through the `fields` parameter or GraphQL if at all possible 🤔  The steps you've described is effectively what the API does internally when using those methods of retrieving nested data 🙂

**Reply by lexoyo** [reply]
Interesting that this is how internals work

> it would be easier to rely on nesting through the fields parameter or GraphQL

Just to be sure I understand, every time the template does `{{ directus.collections.posts[1] | get: "related_pages" }}` I need to

1. call the api, e.g. `GET /items/posts/2?fields=related_pages.*.*` which returns

```json
{
  "data": {
    "related_pages": [
      {
        "id": 1,
        "posts_id": {
          "id": 2,
...same post...
        },
        "page_id": {
...the data I need...
        }
      },
      {
        "id": 2,
        "posts_id": {
          "id": 2,
...same post...
        },
        "page_id": {
...the data I need...
        }
      }
    ]
  }
}
```
3. lookup `meta.junction_field` from `/relations/` to get `page_id`
4. return `data.related_pages` array mapped to `page_id`

Is that right?
It is still pretty complex and involves 1 call for each filter

**Reply by rijkvanzanten** [reply]
Hmm I think I'm misunderstanding exactly what you're trying to achieve.. What do you intent to happen when the user does this?:

> For simple types one could do {{ directus.collections.posts[0] | get: "title" }}

**Reply by lexoyo** [reply]
Thank you @rijkvanzanten for your answer and patience !

Here is another way to put it:

What I want to do is to provide a filter to get an item's related data, e.g `{% assign page = item | directus_relation: 'link' %}`, with `link` being a field on `item` with an alias type. Whatever the type, this filter would return the related item(s)

This filter is just a `function getRelation(item, relationshipFieldName)`

I did not figure out yet

* a better name for this filter
* how to write this filter
* would it be too much calls from 11ty to directus if a template calls the filter 100s of times

I'll keep you posted as I go, but any pointer is welcome and will be full exploited ! ;)
